node/libs/storage/src/block_store.rs

```diff
@@ -0,0 +1,218 @@
+
+    /// Persistently store a block.
+    /// Implementations are only required to accept a block directly after the current last block,
+    /// so that the stored blocks always constitute a continuous range.
+    /// Implementation should return only after the block is stored PERSISTENTLY -
+    /// consensus liveness property depends on this behavior.
+    async fn store_next_block(
+        &self,
+        ctx: &ctx::Ctx,
+        block: &validator::FinalBlock,
+    ) -> ctx::Result<()>;
+}
+
+#[derive(Debug)]
+struct Inner {
+    inmem: sync::watch::Sender<BlockStoreState>,
+    persisted: BlockStoreState,
+    // TODO: this data structure can be optimized to a VecDeque (or even just a cyclical buffer).
+    cache: BTreeMap<validator::BlockNumber, validator::FinalBlock>,
+    cache_capacity: usize,
+}
+
+/// A wrapper around a PersistentBlockStore which adds caching blocks in-memory
+/// and other useful utilities.
+#[derive(Debug)]
+pub struct BlockStore {
+    inner: sync::watch::Sender<Inner>,
+    persistent: Box<dyn PersistentBlockStore>,
+}
+
+/// Runner of the BlockStore background tasks.
+#[must_use]
+pub struct BlockStoreRunner(Arc<BlockStore>);
+
+impl BlockStoreRunner {
+    /// Runs the background tasks of the BlockStore.
+    pub async fn run(self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        let res = async {
+            let inner = &mut self.0.inner.subscribe();
+            loop {
+                let block = sync::wait_for(ctx, inner, |inner| !inner.cache.is_empty())
+                    .await?
+                    .cache
+                    .first_key_value()
+                    .unwrap()
+                    .1
+                    .clone();
+                self.0.persistent.store_next_block(ctx, &block).await?;
+                self.0.inner.send_modify(|inner| {
+                    debug_assert!(inner.persisted.next() == block.header().number);
+                    inner.persisted.last = block.justification.clone();
+                    inner.cache.remove(&block.header().number);
+                });
+            }
+        }
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+}
+
+impl BlockStore {
+    /// Constructs a BlockStore.
+    /// BlockStore takes ownership of the passed PersistentBlockStore,
+    /// i.e. caller should modify the underlying persistent storage
+    /// (add/remove blocks) ONLY through the constructed BlockStore.
+    pub async fn new(
+        ctx: &ctx::Ctx,
+        persistent: Box<dyn PersistentBlockStore>,
+        cache_capacity: usize,
+    ) -> ctx::Result<(Arc<Self>, BlockStoreRunner)> {
+        if cache_capacity < 1 {
+            return Err(anyhow::anyhow!("cache_capacity has to be >=1").into());
+        }
+        let state = persistent
+            .state(ctx)
+            .await?
+            .context("storage empty, expected at least 1 block")?;
```

Comment:
Dumb question: Is this requirement going to be relaxed in the future? Looks weird that on the `PersistentBlockStore` level, empty stores are allowed, but they are not actually supported.

Reply:
bft algorithm implementation assumes that the storage is non-empty, as fetching genesis block from a trusted source is out of scope of bft. I've allowed PersistentBlockStore to be empty to simplify implementations which operate on real storage and need to have a way of inserting a genesis block and a way of checking whether genesis is already there (and it looked weird to me, having special methods all over the place just for that). We can later reconsider whether restricting PersistentBlockStore more would be beneficial.